food_delivery/foodapp/views.py

```diff
@@ -1,3 +1,453 @@
+    """
+    Handles custom actions for specific orders: calculating delivery time, marking as delivered, canceling.
+    """
+
+    def get_object(self, pk):
+        try:
+            return Order.objects.get(pk=pk)
+        except Order.DoesNotExist:
+            return None
+
+    def calculate_delivery_time(self, order):
+        """
+        A function to calculate the estimated delivery time for an order.
+        It uses the geolocation of the restaurant and the customer.
+        """
+        geolocator = Nominatim(user_agent="food_delivery_app")
+        restaurant_location = geolocator.geocode(order.restaurant.address)
+        customer_location = geolocator.geocode(order.customer.current_location)
+
+        if restaurant_location and customer_location:
+            restaurant_coords = (restaurant_location.latitude, restaurant_location.longitude)
+            customer_coords = (customer_location.latitude, customer_location.longitude)
+            distance = geodesic(restaurant_coords, customer_coords).km
+            
+            # Assuming average delivery speed is 40 km/h
+            estimated_time = distance / 40 * 60  # Time in minutes
+            return datetime.now() + timedelta(minutes=estimated_time)
+        return None
+
+    def get(self, request, pk, action):
+        """
+        GET method for actions like 'estimated_time'.
+        Accessible at: /orders/{pk}/actions/{action}/ (e.g., /orders/1/actions/estimated_time/)
+        """
+        order = self.get_object(pk)
+        if order is None:
+            return Response({"error": "Order not found"}, status=status.HTTP_404_NOT_FOUND)
+
+        if action == 'estimated_time':
+            estimated_time = self.calculate_delivery_time(order)
+            if estimated_time:
+                return Response({"estimated_delivery_time": estimated_time.strftime("%Y-%m-%d %H:%M:%S")}, status=status.HTTP_200_OK)
+            return Response({"error": "Could not calculate delivery time"}, status=status.HTTP_400_BAD_REQUEST)
+
+        return Response({"error": "Invalid action"}, status=status.HTTP_400_BAD_REQUEST)
+
+    def post(self, request, pk, action):
+        """
+        POST method for actions like 'mark_as_delivered' and 'cancel_order'.
+        Accessible at: /orders/{pk}/actions/{action}/ (e.g., /orders/1/actions/mark_as_delivered/)
+        """
+        order = self.get_object(pk)
+        if order is None:
+            return Response({"error": "Order not found"}, status=status.HTTP_404_NOT_FOUND)
+
+        if action == 'mark_as_delivered':
+            if order.status == 'DELIVERED':
+                return Response({"message": "Order already delivered"}, status=status.HTTP_400_BAD_REQUEST)
+            order.status = 'DELIVERED'
+            order.save()
+            return Response({"message": "Order marked as delivered"}, status=status.HTTP_200_OK)
+
+        elif action == 'cancel_order':
+            if order.status in ['DELIVERED', 'CANCELLED']:
```

Comment:
use django models TextChoices for these statuses or any kind of static keywords. don't use string everywhere.

Reply:
Done sir.Used TextChoices